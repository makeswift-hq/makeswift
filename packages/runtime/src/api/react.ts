import * as MakeswiftApiClient from '../state/makeswift-api-client'
import {
  APIResourceType,
  File,
  GlobalElement,
  LocalizedGlobalElement,
  Page,
  PagePathnameSlice,
  Site,
  Snippet,
  Swatch,
  Table,
  Typography,
} from './graphql/types'
import { GraphQLClient } from './graphql/client'
import { CreateTableRecordMutation } from './graphql/documents'
import {
  CreateTableRecordMutationResult,
  CreateTableRecordMutationVariables,
} from './graphql/generated/types'

export type SerializedLocalizedResourcesMap = {
  [resourceId: string]: string | null
}

type LocalizedResourcesMap = Map<string, string | null>

export type CacheData = MakeswiftApiClient.SerializedState

export type MakeswiftClientOptions = {
  uri: string
  cacheData?: CacheData
  localizedResourcesMap?: SerializedLocalizedResourcesMap
  locale?: Intl.Locale
}

/**
 * NOTE(miguel): This "client" is used to fetch Makeswift API resources needed for the host. For
 * example, swatches, files, typographies, etc. Ideally it's internal to the runtime and is only
 * used by controls to transform API references to API resources.
 *
 * Moreover, it's use should be reserved for the builder only, since for live pages all Makeswift
 * API resources should be embedded in the "page snapshot". In the builder, this client serves the
 * purpose of sending requests for API resources and keeping a cache so that changes that happen in
 * the builder, like modifying a swatch, can be sent via `postMessage` to the host and the cache can
 * immediately update the value and re-render.
 *
 * Furthermore, the API resources requested shouldn't be requested directly from the Makeswift API
 * as that would require those resources to not be authenticated since the requests come from the
 * browser when running the host. Instead, the requests should go to the host directly, at the
 * Makeswift API endpoint (i.g., `/api/makeswift/[...makeswift]` dynamic route) where the host's
 * API key can be used, securely, in the server. For this reason, this client should really be a
 * client of the host's API, not Makeswift's, intended to build and continuously maintain a realtime
 * snapshot for use in the builder, not the lives pages.
 */
export class MakeswiftHostApiClient {
  graphqlClient: GraphQLClient
  makeswiftApiClient: MakeswiftApiClient.Store
  subscribe: MakeswiftApiClient.Store['subscribe']
  private localizedResourcesMap: LocalizedResourcesMap
  private locale: Intl.Locale | undefined

  constructor({ uri, cacheData, localizedResourcesMap = {}, locale }: MakeswiftClientOptions) {
    this.graphqlClient = new GraphQLClient(uri)
    this.makeswiftApiClient = MakeswiftApiClient.configureStore({ serializedState: cacheData })
    this.subscribe = this.makeswiftApiClient.subscribe
    this.localizedResourcesMap = new Map(Object.entries(localizedResourcesMap))
    this.locale = locale
  }

  readSwatch(swatchId: string): Swatch | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Swatch,
      swatchId,
    )
  }

  async fetchSwatch(swatchId: string): Promise<Swatch | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Swatch, swatchId),
    )
  }

  readFile(fileId: string): File | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.File,
      fileId,
    )
  }

  async fetchFile(fileId: string): Promise<File | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.File, fileId),
    )
  }

  readTypography(typographyId: string): Typography | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Typography,
      typographyId,
    )
  }

  async fetchTypography(typographyId: string): Promise<Typography | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Typography, typographyId),
    )
  }

  readGlobalElement(globalElementId: string): GlobalElement | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.GlobalElement,
      globalElementId,
    )
  }

  async fetchGlobalElement(globalElementId: string): Promise<GlobalElement | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.GlobalElement, globalElementId),
    )
  }

  readLocalizedGlobalElement(globalElementId: string): LocalizedGlobalElement | null {
    const localizedGlobalElementId = this.getLocalizedResourceId(globalElementId)

    if (localizedGlobalElementId == null) return null

    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.LocalizedGlobalElement,
      localizedGlobalElementId,
    )
  }

  async fetchLocalizedGlobalElement(
    globalElementId: string,
  ): Promise<LocalizedGlobalElement | null> {
    const locale = this.locale

    if (locale == null) return null

    // If getLocalizedResourceId returns null, it means we have tried to fetch the resource,
    // but the resource is not available. If we haven't fetched it yet, it'll return undefined.
    const noLocalizedResource = this.getLocalizedResourceId(globalElementId) === null

    if (noLocalizedResource) return null

    // TODO(fikri): We're checking the cache here because the cache hit will fail on fetchAPIResource.
    // This is because in the cache we're saving the ID of the localizedGlobalElementId,
    // but we're reading the cache using the globalElementId.
    // This is very confusing and it can lead to more bugs. We need to refactor how we handle
    // localized global element.
    const cacheResult = this.readLocalizedGlobalElement(globalElementId)

    if (cacheResult) return cacheResult

    const result = await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(
        APIResourceType.LocalizedGlobalElement,
        globalElementId,
        locale,
      ),
    )

    this.setLocalizedResourceId({
      resourceId: globalElementId,
      localizedResourceId: result?.id ?? null,
    })

    return result
  }

  readPagePathnameSlice(pageId: string): PagePathnameSlice | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.PagePathnameSlice,
      pageId,
    )
  }

  async fetchPagePathnameSlice(pageId: string): Promise<PagePathnameSlice | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.PagePathnameSlice, pageId, this.locale),
    )
  }

  readTable(tableId: string): Table | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Table,
      tableId,
    )
  }

  async fetchTable(tableId: string): Promise<Table | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Table, tableId),
    )
  }

  async createTableRecord(tableId: string, columns: any): Promise<void> {
    await this.graphqlClient.request<
      CreateTableRecordMutationResult,
      CreateTableRecordMutationVariables
    >(CreateTableRecordMutation, { input: { data: { tableId, columns } } })
  }

  readSite(siteId: string): Site | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Site,
      siteId,
    )
  }

  readPage(pageId: string): Page | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Page,
      pageId,
    )
  }

  readSnippet(snippetId: string): Snippet | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Snippet,
      snippetId,
    )
  }

  private getLocalizedResourceId(resourceId: string): string | undefined | null {
    return this.localizedResourcesMap?.get(resourceId)
  }

  setLocalizedResourceId({
    resourceId,
    localizedResourceId,
  }: {
    resourceId: string
    localizedResourceId: string | null
  }): void {
    this.localizedResourcesMap.set(resourceId, localizedResourceId)
  }
}
