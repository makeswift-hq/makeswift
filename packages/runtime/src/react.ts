export type { Action } from './state/actions'
export { ActionTypes, changeDocument } from './state/actions'
export {
  Document,
  Element,
  RuntimeProvider,
  ReactRuntime,
  useIsInBuilder,
  PageProvider,
  usePageId,
} from './runtimes/react'
