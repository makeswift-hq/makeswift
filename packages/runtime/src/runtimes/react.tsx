import {
  createContext,
  forwardRef,
  memo,
  ReactNode,
  Ref,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react'
import { useSyncExternalStoreWithSelector } from 'use-sync-external-store/shim/with-selector'

import * as ReactPage from '../state/react-page'
import type * as ReactBuilderPreview from '../state/react-builder-preview'
import {
  changeComponentHandle,
  mountComponentEffect,
  registerComponentEffect,
  registerReactComponentEffect,
} from '../state/actions'
import type {
  PropControllerDescriptor,
  PropControllerDescriptorValueType,
} from '../prop-controllers'
import { ComponentIcon } from '../state/modules/components-meta'

const contextDefaultValue = ReactPage.configureStore()

export const ReactRuntime = {
  registerComponent<
    P extends Record<string, PropControllerDescriptor>,
    C extends ReactPage.ComponentType<{ [K in keyof P]?: PropControllerDescriptorValueType<P[K]> }>,
  >(
    component: C,
    {
      type,
      label,
      icon = 'Cube40',
      hidden = false,
      props,
    }: { type: string; label: string; icon?: ComponentIcon; hidden?: boolean; props?: P },
  ): () => void {
    const unregisterComponent = contextDefaultValue.dispatch(
      registerComponentEffect(type, { label, icon, hidden }, props ?? {}),
    )

    const unregisterReactComponent = contextDefaultValue.dispatch(
      registerReactComponentEffect(type, component as unknown as ReactPage.ComponentType),
    )

    return () => {
      unregisterComponent()
      unregisterReactComponent()
    }
  },
}

const Context = createContext(contextDefaultValue)

type RuntimeProviderProps = {
  defaultRootElements?: Map<string, ReactPage.Element>
  children?: ReactNode
}

export function RuntimeProvider({
  children,
  defaultRootElements,
}: RuntimeProviderProps): JSX.Element {
  const [store, setStore] = useState(() =>
    ReactPage.configureStore({
      preloadedState: contextDefaultValue.getState(),
      rootElements: defaultRootElements,
    }),
  )

  useEffect(() => {
    // TODO(miguel): perform a more robust validation.
    const isInBuilder = window.parent !== window

    if (!isInBuilder) return

    const initializePromise = initializeReactBuilderPreview()

    return () => {
      initializePromise.then(cleanUp => {
        cleanUp()
      })
    }

    async function initializeReactBuilderPreview(): Promise<() => void> {
      const ReactBuilderPreview = await import('../state/react-builder-preview')

      const store = await new Promise<ReactBuilderPreview.Store>(resolve => {
        setStore(store => {
          const nextStore = ReactBuilderPreview.configureStore({ preloadedState: store.getState() })

          resolve(nextStore)

          return nextStore
        })
      })

      return store.dispatch(ReactBuilderPreview.initialize())
    }
  }, [])

  return <Context.Provider value={store}>{children}</Context.Provider>
}

const DocumentContext = createContext<string | null>(null)

function useDocumentKey(): string | null {
  return useContext(DocumentContext)
}

type State = ReactPage.State | ReactBuilderPreview.State

function useSelector<R>(selector: (state: State) => R): R {
  const store = useContext(Context)

  return useSyncExternalStoreWithSelector(store.subscribe, store.getState, store.getState, selector)
}

function useComponent(type: string): ReactPage.ComponentType | null {
  return useSelector(state => ReactPage.getReactComponent(state, type))
}

export function useElementId(elementKey: string | null | undefined): string | null {
  const documentKey = useDocumentKey()

  return useSelector(state =>
    documentKey == null || elementKey == null
      ? null
      : ReactPage.getElementId(state, documentKey, elementKey),
  )
}

function useDocumentRootElement(documentKey: string): ReactPage.Element | null {
  return useSelector(state => ReactPage.getDocumentRootElement(state, documentKey))
}

export function useIsInBuilder(): boolean {
  return useSelector(state => ReactPage.getIsInBuilder(state))
}

type Dispatch = ReactPage.Dispatch & ReactBuilderPreview.Dispatch

function useDispatch(): Dispatch {
  const store = useContext(Context)

  return store.dispatch
}

type ElementDataProps = {
  elementData: ReactPage.ElementData
}

const ElementData = memo(
  forwardRef(function ElementData(
    { elementData }: ElementDataProps,
    ref: Ref<unknown>,
  ): JSX.Element {
    const Component = useComponent(elementData.type)

    if (Component == null) {
      return (
        <div ref={ref as Ref<HTMLDivElement>}>
          <p>Component Not Found</p>
          <pre>
            <code>{JSON.stringify(elementData, null, 2)}</code>
          </pre>
        </div>
      )
    }

    return <Component {...elementData.props} key={elementData.key} ref={ref} />
  }),
)

type ElementRefereceProps = {
  elementReference: ReactPage.ElementReference
}

const ElementReference = memo(
  forwardRef(function ElementReference(
    { elementReference }: ElementRefereceProps,
    ref: Ref<unknown>,
  ): JSX.Element {
    return (
      <div ref={ref as Ref<HTMLDivElement>}>
        <p>Not Implemented</p>
        <pre>
          <code>{JSON.stringify(elementReference, null, 2)}</code>
        </pre>
      </div>
    )
  }),
)

type ElementProps = {
  element: ReactPage.Element
}

export const Element = memo(function Element({ element }: ElementProps): JSX.Element {
  const elementKey = element.key
  const dispatch = useDispatch()
  const documentKey = useDocumentKey()
  const ref = useCallback(
    (handle: unknown): void => {
      if (documentKey) dispatch(changeComponentHandle(documentKey, elementKey, handle))
    },
    [dispatch, elementKey],
  )

  useEffect(() => {
    if (documentKey == null) return

    return dispatch(mountComponentEffect(documentKey, elementKey))
  }, [dispatch, documentKey, elementKey])

  return ReactPage.isElementReference(element) ? (
    <ElementReference key={elementKey} ref={ref} elementReference={element} />
  ) : (
    <ElementData key={elementKey} ref={ref} elementData={element} />
  )
})

type DocumentProps = {
  documentKey: string
}

export const Document = memo(function Document({ documentKey }: DocumentProps): JSX.Element {
  const documentRootElement = useDocumentRootElement(documentKey)

  if (documentRootElement == null) {
    return (
      <div>
        <p>Document Not Found</p>
        <pre>
          <code>{JSON.stringify({ documentKey }, null, 2)}</code>
        </pre>
      </div>
    )
  }

  return (
    <DocumentContext.Provider value={documentKey}>
      <Element element={documentRootElement} />
    </DocumentContext.Provider>
  )
})
