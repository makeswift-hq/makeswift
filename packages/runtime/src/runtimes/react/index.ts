export { ReactRuntime } from './react-runtime'

export { Element } from './components/Element'
export { DocumentReference } from './components/DocumentReference'
export { RuntimeProvider } from './components/RuntimeProvider'
export { PageProvider } from './components/PageProvider'

export { useBreakpoints } from './hooks/use-breakpoints'
export { useBuilderEditMode } from './hooks/use-builder-edit-mode'
export { useDocumentKey } from './hooks/use-document-key'
export { useElementId } from './hooks/use-element-id'
export { useIsInBuilder } from './hooks/use-is-in-builder'
export { useIsPreview } from './hooks/use-is-preview'
export { usePageId } from './hooks/use-page-id'
export { useSelector } from './hooks/use-selector'
export { StoreContext, useStore } from './hooks/use-store'
