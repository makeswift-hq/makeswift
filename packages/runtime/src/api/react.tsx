import {
  ApolloClient,
  DocumentNode,
  NormalizedCacheObject,
  OperationVariables,
  QueryHookOptions,
  QueryResult,
  TypedDocumentNode,
  useQuery as useApolloQuery,
  useMutation as useApolloMutation,
  DefaultContext,
  ApolloCache,
  MutationHookOptions,
  MutationTuple,
  InMemoryCache,
  TypePolicies,
} from '@apollo/client'
import { getDataFromTree } from '@apollo/client/react/ssr'
export { gql } from '@apollo/client'
import { createContext, ReactNode, useContext } from 'react'
import uuid from 'uuid/v4'

import { DocumentReference, RuntimeProvider } from '../react'
import { createDocumentReference, Element } from '../state/react-page'

const typePolicies: TypePolicies = {
  Query: {
    fields: {
      swatches(existingData, { args, toReference }) {
        return (
          existingData ??
          args?.ids.map((id: string) => toReference({ __typename: 'Swatch', id }, true))
        )
      },
      file(existingData, { args, toReference }) {
        return existingData ?? toReference({ __typename: 'File', id: args?.id }, true)
      },
      files(existingData, { args, toReference }) {
        return (
          existingData ??
          args?.ids.map((id: string) => toReference({ __typename: 'File', id }, true))
        )
      },
      typographies(existingData, { args, toReference }) {
        return (
          existingData ??
          args?.ids.map((id: string) => toReference({ __typename: 'Typography', id }, true))
        )
      },
      pagePathnamesById(existingData, { args, toReference }) {
        return (
          existingData ??
          args?.ids.map((id: string) => toReference({ __typename: 'PagePathnameSlice', id }, true))
        )
      },
      globalElement(existingData, { args, toReference }) {
        return existingData ?? toReference({ __typename: 'GlobalElement', id: args?.id }, true)
      },
      table(existingData, { args, toReference }) {
        return existingData ?? toReference({ __typename: 'Table', id: args?.id }, true)
      },
    },
  },
}

type CreateApolloClientParams = {
  uri?: string
  cacheData?: NormalizedCacheObject
}

export function createApolloClient({ uri, cacheData }: CreateApolloClientParams) {
  const cache = new InMemoryCache({ typePolicies })

  if (cacheData) cache.restore(cacheData)

  return new ApolloClient({ uri, cache })
}

export type MakeswiftClientOptions = {
  uri?: string
  cacheData?: NormalizedCacheObject
}

export class MakeswiftClient {
  apolloClient: ApolloClient<NormalizedCacheObject>

  constructor({ uri, cacheData }: MakeswiftClientOptions) {
    this.apolloClient = createApolloClient({ uri, cacheData })
  }

  async prefetch(element: Element): Promise<NormalizedCacheObject> {
    const id = uuid()

    await getDataFromTree(
      <RuntimeProvider client={this} defaultRootElements={new Map([[id, element]])}>
        <DocumentReference documentReference={createDocumentReference(id)} />
      </RuntimeProvider>,
    )

    return this.apolloClient.cache.extract()
  }
}

const Context = createContext<MakeswiftClient | undefined>(undefined)

export function useQuery<TData = any, TVariables = OperationVariables>(
  query: DocumentNode | TypedDocumentNode<TData, TVariables>,
  options?: QueryHookOptions<TData, TVariables>,
): QueryResult<TData, TVariables> {
  const client = useContext(Context)

  return useApolloQuery(query, { client: client?.apolloClient, ...options })
}

export function useMutation<
  TData = any,
  TVariables = OperationVariables,
  TContext = DefaultContext,
  TCache extends ApolloCache<any> = ApolloCache<any>,
>(
  mutation: DocumentNode | TypedDocumentNode<TData, TVariables>,
  options?: MutationHookOptions<TData, TVariables, TContext>,
): MutationTuple<TData, TVariables, TContext, TCache> {
  const client = useContext(Context)

  return useApolloMutation(mutation, { client: client?.apolloClient, ...options })
}

type MakeswiftProviderProps = {
  client: MakeswiftClient
  children: ReactNode
}

export function MakeswiftProvider({ client, children }: MakeswiftProviderProps) {
  return <Context.Provider value={client}>{children}</Context.Provider>
}
