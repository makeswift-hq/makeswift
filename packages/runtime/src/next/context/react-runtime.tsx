'use client'

import { ReactNode, createContext, useContext } from 'react'
import { ReactRuntime } from '../../react'

const Context = createContext<ReactRuntime>(ReactRuntime)

export function useReactRuntime(): ReactRuntime {
  return useContext(Context)
}

export function ReactRuntimeProvider({
  children,
  runtime,
}: {
  children: ReactNode
  runtime: ReactRuntime
}) {
  return <Context.Provider value={runtime}>{children}</Context.Provider>
}
