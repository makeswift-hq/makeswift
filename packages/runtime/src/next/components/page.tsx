'use client'

import { Suspense, memo, useMemo } from 'react'

import { RuntimeProvider } from '../../runtimes/react'
import { Page as PageMeta } from '../../components/page'
import { MakeswiftHostApiClient } from '../../api/react'
import { MakeswiftPageSnapshot } from '../client'

export type PageProps = {
  snapshot: MakeswiftPageSnapshot
}

export const Page = memo(({ snapshot }: PageProps) => {
  const client = useMemo(
    () =>
      new MakeswiftHostApiClient({
        uri: new URL('graphql', snapshot.apiOrigin).href,
        cacheData: snapshot.cacheData,
        localizedResourcesMap: snapshot.localizedResourcesMap,
        locale: snapshot.locale ? new Intl.Locale(snapshot.locale) : undefined,
      }),
    [snapshot],
  )
  const rootElements = new Map([[snapshot.document.id, snapshot.document.data]])

  snapshot.document.localizedPages.forEach(localizedPage => {
    rootElements.set(localizedPage.elementTreeId, localizedPage.data)
  })

  return (
    <Suspense>
      <RuntimeProvider client={client} rootElements={rootElements} preview={snapshot.preview}>
        {/* We use a key here to reset the Snippets state in the PageMeta component */}
        <PageMeta key={snapshot.document.data.key} document={snapshot.document} />
      </RuntimeProvider>
    </Suspense>
  )
})
