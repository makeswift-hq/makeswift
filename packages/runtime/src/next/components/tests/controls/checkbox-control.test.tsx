/** @jest-environment jsdom */

import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import { ElementData } from '../../../../state/react-page'
import { Page } from '../../page'
import { act } from 'react-dom/test-utils'
import { ReactRuntimeProvider } from '../../../context/react-runtime'
import { ReactRuntime } from '../../../../react'
import { forwardRef } from 'react'
import {
  createMakeswiftPageSnapshot,
  createRootComponent,
} from '../../../../utils/tests/element-data-test-test'
import {
  Checkbox,
  CheckboxControlData,
  CheckboxControlDataTypeKey,
  CheckboxControlDataTypeValueV1,
  CheckboxControlDataV0,
  CheckboxControlDataV1,
} from '@makeswift/controls'

const ROOT_ID = '00000000-0000-0000-0000-000000000000'
const ELEMENT_ID = '11111111-1111-1111-1111-111111111111'

describe('Page', () => {
  test.each([true, false])('can render CheckboxControl v0 data', async bool => {
    // Arrange
    const checkboxControlData: CheckboxControlDataV0 = bool
    const TestComponentType = 'TestComponent'
    const testId = 'test-id'
    const elementData: ElementData = createRootComponent(
      [
        {
          key: ELEMENT_ID,
          type: TestComponentType,
          props: {
            checkbox: checkboxControlData,
          },
        },
      ],
      ROOT_ID,
    )
    const snapshot = createMakeswiftPageSnapshot(elementData)
    const runtime = new ReactRuntime()

    // Act
    runtime.registerComponent(
      forwardRef<HTMLDivElement, { checkbox?: CheckboxControlData }>(({ checkbox }, ref) => {
        return (
          <div ref={ref} data-testid={testId}>
            {JSON.stringify(checkbox)}
          </div>
        )
      }),
      {
        type: TestComponentType,
        label: 'TestComponent',
        props: {
          checkbox: Checkbox(),
        },
      },
    )

    // Assert
    await act(async () =>
      render(
        <ReactRuntimeProvider runtime={runtime}>
          <Page snapshot={snapshot} />
        </ReactRuntimeProvider>,
      ),
    )

    expect(snapshot).toMatchSnapshot()
    expect(JSON.parse(screen.getByTestId(testId).textContent ?? '')).toBe(bool)
  })

  test.each([true, false])('can render CheckboxControl v1 data', async bool => {
    // Arrange
    const checkboxControlData: CheckboxControlDataV1 = {
      [CheckboxControlDataTypeKey]: CheckboxControlDataTypeValueV1,
      value: bool,
    }

    const TestComponentType = 'TestComponent'
    const testId = 'test-id'
    const elementData: ElementData = createRootComponent(
      [
        {
          key: ELEMENT_ID,
          type: TestComponentType,
          props: {
            checkbox: checkboxControlData,
          },
        },
      ],
      ROOT_ID,
    )
    const snapshot = createMakeswiftPageSnapshot(elementData)
    const runtime = new ReactRuntime()

    // Act
    runtime.registerComponent(
      forwardRef<HTMLDivElement, { checkbox?: CheckboxControlData }>(({ checkbox }, ref) => {
        return (
          <div ref={ref} data-testid={testId}>
            {JSON.stringify(checkbox)}
          </div>
        )
      }),
      {
        type: TestComponentType,
        label: 'TestComponent',
        props: {
          checkbox: Checkbox(),
        },
      },
    )

    // Assert
    await act(async () =>
      render(
        <ReactRuntimeProvider runtime={runtime}>
          <Page snapshot={snapshot} />
        </ReactRuntimeProvider>,
      ),
    )

    expect(snapshot).toMatchSnapshot()
    expect(JSON.parse(screen.getByTestId(testId).textContent ?? '')).toBe(bool)
  })
})
