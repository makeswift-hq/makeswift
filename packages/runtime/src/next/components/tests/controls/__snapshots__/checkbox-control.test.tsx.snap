// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page can render CheckboxControl v0 data 1`] = `
{
  "apiOrigin": "https://test-api-origin.com",
  "cacheData": {},
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "checkbox": true,
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
  "locale": null,
  "localizedResourcesMap": {},
  "preview": false,
}
`;

exports[`Page can render CheckboxControl v1 data 1`] = `
{
  "apiOrigin": "https://test-api-origin.com",
  "cacheData": {},
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "checkbox": {
                  "@@makeswift/type": "checkbox::v1",
                  "value": true,
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
  "locale": null,
  "localizedResourcesMap": {},
  "preview": false,
}
`;
